pubnub-kotlin/src/main/kotlin/com/pubnub/api/v2/PNConfiguration.kt

```diff
@@ -0,0 +1,76 @@
+package com.pubnub.api.v2
+
+import com.pubnub.api.UserId
+import com.pubnub.api.crypto.CryptoModule
+import com.pubnub.api.enums.PNHeartbeatNotificationOptions
+import com.pubnub.api.enums.PNLogVerbosity
+import com.pubnub.api.retry.RetryConfiguration
+import com.pubnub.internal.v2.BasePNConfigurationImpl
+import com.pubnub.internal.v2.PNConfigurationImpl
+import okhttp3.Authenticator
+import okhttp3.CertificatePinner
+import okhttp3.ConnectionSpec
+import okhttp3.logging.HttpLoggingInterceptor
+import java.net.Proxy
+import java.net.ProxySelector
+import javax.net.ssl.HostnameVerifier
+import javax.net.ssl.SSLSocketFactory
+import javax.net.ssl.X509ExtendedTrustManager
+
+interface PNConfiguration : BasePNConfiguration {
```

Comment:
In kotlin there are "named parameters". I thinking do we need builder patter for kotlin. What advantages has this approach over constructor with "named parameters"?

Reply:
there are still advantages sometimes, especially for long parameter lists:
* it allows the user to pass around the builder or gradually build up the required parameters, without having to store them in a long list of variables to then call the constructor once
* it lets us start building a small DSL for when we start introducing groups of parameters (e.g. httpOptions, keyOptions, cryptoOptions) - tbf this could be also done in the constructor, but becomes a bit unwieldy with nested calls...
* it lets us evolve the available options and have setters with custom logic, for example making a change to a constructor is usually a breaking change (unless you're adding a new option at the end of the parameter list), but deprecating an option and adding a new one is useful to have

tl;dr it gives both us and the user more flexibility in how to construct the final object

Reply:
ok, thanks